image-loader/app/controllers/ImageLoaderController.scala

```diff
@@ -43,207 +38,130 @@ class ImageLoaderController(auth: Authentication, downloader: Downloader, store:
+    // synchronous write to file
+    val tempFile = createTempFile("requestBody")
+    Logger.info("body parsed")
+    val parsedBody = DigestBodyParser.create(tempFile)
 
+    // asynchronous load, store and delete
     auth.async(parsedBody) { req =>
-      val result = loadFile(req.body, req.user, uploadedBy, identifiers, DateTimeUtils.fromValueOrNow(uploadTime), filename, requestContext)
-      Logger.info("loadImage request end")(requestContext.toMarker(markers))
-      result
+      val result = for {
+        uploadRequest <- imageImporter.loadFile(
+          req.body,
+          req.user,
+          uploadedBy,
+          identifiers,
+          DateTimeUtils.fromValueOrNow(uploadTime),
+          filename.flatMap(_.trim.nonEmptyOpt),
+          context)
+        result <- imageImporter.storeFile(uploadRequest)
+      } yield result
+
+      result.onComplete( _ => Try { deleteTempFile(tempFile) } )
+
+      result map { r =>
+        Ok(r)
+      } recover {
+        case e: ImageLoaderException => InternalServerError(Json.obj("error" -> e.getMessage))
+      }
     }
   }
 
-  private def getSrcFileDigestForProjection(s3Src: S3Object, imageId: String, requestLoggingContext: RequestLoggingContext) = {
-    val uploadedFile = createTempFile(s"projection-${imageId}", requestLoggingContext)
-    IOUtils.copy(s3Src.getObjectContent, new FileOutputStream(uploadedFile))
-    DigestedFile(uploadedFile, imageId)
-  }
-
-  def projectImageBy(imageId: String) = {
-    val requestContext = RequestLoggingContext(
+  // Fetch
+  def projectImageBy(imageId: String): Action[AnyContent] = {
+    implicit val requestContext: RequestLoggingContext = RequestLoggingContext(
       initialMarkers = Map(
         "imageId" -> imageId,
         "requestType" -> "image-projection"
       )
     )
 
-    auth { _ =>
-      projectS3ImageById(imageId, requestContext) match {
-        case Success(maybeImage) =>
-          maybeImage match {
-            case Some(img) => {
-              Logger.info("image found")(requestContext.toMarker())
-              Ok(Json.toJson(img)).as(ArgoMediaType)
-            }
-            case None =>
-              val s3Path = "s3://" + config.imageBucket + "/" + ImageIngestOperations.fileKeyFromId(imageId)
-              Logger.info("image not found")(requestContext.toMarker())
-              respondError(NotFound, "image-not-found", s"Could not find image: $imageId in s3 at $s3Path")
-          }
-        case Failure(error) => {
-          Logger.error(s"image projection failed", error)(requestContext.toMarker())
-          respondError(InternalServerError, "image-projection-failed", error.getMessage)
-        }
+    val tempFile = createTempFile(s"projection-$imageId")
+    Logger.info("XXX" )
+    auth.async { _ =>
+      val result = for {
+        result <- imageProjector.projectS3ImageById(imageUploadProjector, imageId, requestContext, tempFile)
+      } yield result
```

Comment:
Correct me if wrong, but I think this desugars to 

```
imageProjector.projectS3ImageById(imageUploadProjector, imageId, requestContext, tempFile).map(identity)
```

that is to say, we can write this more simply as

```
val result = imageProjector.projectS3ImageById(imageUploadProjector, imageId, requestContext, tempFile)
```